### PR TITLE
Issue 8 add request controllers

### PR DESCRIPTION
**Pull request by JunZheng007**
### What this PR does (required):
- GET: list of requests for logged in user
- POST: Create a new request
- UPDATE: Update request with approved or decline

### Any information needed to test this feature (required):
- need to user request model based on issue #9(PR #56)

### CODE REVIEW

**Comment by bonnieli** [comment]
Hey @JunZheng007 Looks like there's a merge conflict here to resolve :)

File: server/controllers/request.js [hunk 1]
```diff
@@ -0,0 +1,73 @@
+const Request = require("../models/Request");
+const Profile = require("../models/Profile");
+const asyncHandler = require("express-async-handler");
+
+// @route GET /request
+// @desc list of requests
+// @access Private
```
Comment (ezeilo-su) [inline server/controllers/request.js:7]: Good documentation!

File: server/controllers/request.js [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+exports.getRequests = asyncHandler(async (req, res, next) => {
+  const profile = await Profile.findOne({ userId: req.user.id });
+
+  if (profile.accountType === "pet_owner") {
+    const requests = await Request.find({ userId: req.user.id });
+    console.log(requests);
```
Comment (ezeilo-su) [inline server/controllers/request.js:13]: Can we get rid of this log to the console?

File: server/routes/request.js [hunk 3]
```diff
@@ -0,0 +1,14 @@
...
+  getRequests,
+  makeRequest,
+  updateRequest,
+} = require("../controllers/request");
+const protect = require("../middleware/auth");
+const router = express.Router();
+
+router.route("/").get(protect, getRequests);
+router.route("/").post(protect, makeRequest);
+router.route("/").put(protect, updateRequest);
```
Comment (ezeilo-su) [inline server/routes/request.js:12]: It's a good practice to pass the id of the resource to be updated in the request params. You can make this one a dynamic route.

**Review by ezeilo-su — COMMENTED** [review]
Hey @JunZheng007, Good work on this one.

File: server/controllers/request.js [hunk 4]
```diff
@@ -0,0 +1,73 @@
...
+      },
+    });
+  }
+});
+
+// @route POST /request
+// @desc Create a new request
+// @access Private
+exports.makeRequest = asyncHandler(async (req, res, next) => {
+  const request = new Request(req.body.request);
```
Comment (rajivtitus) [inline server/controllers/request.js:33]: It is recommended that we destructure the request body for security purposes (to only allow selected fields)

File: server/controllers/request.js [hunk 5]
```diff
@@ -0,0 +1,73 @@
...
+    throw new Error("sitter doesn't exist");
+  }
+
+  if (!request.startDate || !request.endDate) {
+    res.status(400);
+    throw new Error("date input error");
+  }
+
+  await Request.create(request);
+  res.status(200).json("create profile success");
```
Comment (rajivtitus) [inline server/controllers/request.js:51]: Let's also send the newly created request doc with the response

File: server/controllers/request.js [hunk 6]
```diff
@@ -0,0 +1,73 @@
...
+  }
+});
+
+// @route POST /request
+// @desc Create a new request
+// @access Private
+exports.makeRequest = asyncHandler(async (req, res, next) => {
+  const request = new Request(req.body.request);
+
+  if (!request.userId) {
```
Comment (rajivtitus) [inline server/controllers/request.js:35]: Nice to see validation! 

Optional: Alternatively, we can move this into a validation middleware to cleanup the code and increase maintainability

File: server/controllers/request.js [hunk 7]
```diff
@@ -0,0 +1,73 @@
...
+
+  await Request.create(request);
+  res.status(200).json("create profile success");
+});
+
+// @route PUT /request
+// @desc Update request with approved or decline
+// @access Private
+exports.updateRequest = asyncHandler(async (req, res, next) => {
+  const request = await Request.findById(req.params.id);
```
Comment (rajivtitus) [inline server/controllers/request.js:58]: Only the sitter who owns the request should be able to update it

File: server/routes/request.js [hunk 8]
```diff
@@ -0,0 +1,14 @@
+const express = require("express");
+const {
+  getRequests,
+  makeRequest,
+  updateRequest,
+} = require("../controllers/request");
+const protect = require("../middleware/auth");
+const router = express.Router();
+
+router.route("/").get(protect, getRequests);
```
Comment (rajivtitus) [inline server/routes/request.js:10]: Nice routes 👍🏽

File: server/routes/request.js [hunk 9]
```diff
@@ -0,0 +1,14 @@
...
+router.route("/").post(protect, makeRequest);
+router.route("/:id").put(protect, updateRequest);
```
Comment (rajivtitus) [inline server/routes/request.js:12]:
```suggestion
router.route("/:id").patch(protect, updateRequest);
```
If we are only updating part of the request doc, the PATCH HTTP verb seems more appropriate

**Review by rajivtitus — COMMENTED** [review]
Good work Jun! Some comments for you to have a look and address, please let me know if you have any questions

**Comment by JunZheng007** [comment]
all changes merged to PR #70

**Pull request closed without merging**